### PR TITLE
Bluetooth: Mesh: Fail init on model init error

### PR DESCRIPTION
**Pull request by trond-snekvik**
Adds propagation of error returns from the model init callbacks in
Access, and removing any other checks for successful init in the
foundation models.

Signed-off-by: Trond Einar Snekvik <Trond.Einar.Snekvik@nordicsemi.no>

### CODE REVIEW

**Review by jhedberg — APPROVED** [review]

**Comment by jhedberg** [comment]
I guess in some ways this could be seen as fixing  #23177 as well. I wouldn't mind including it into 2.3, but ultimately it's up to @carlescufi

**Review by carlescufi — APPROVED** [review]

File: subsys/bluetooth/mesh/access.c
Comment (Kupeo) [inline subsys/bluetooth/mesh/access.c:309]: https://github.com/zephyrproject-rtos/zephyr/blob/d2c74eb98754e824196541cc6fa6a420fb4ed611/subsys/bluetooth/mesh/access.c#L340-L348

should add check `*err is NULL` ???

Comment (jhedberg) [reply subsys/bluetooth/mesh/access.c:309]: Either that or pass an err to the unprovision case

**Review by jhedberg — CHANGES_REQUESTED** [review]
Potential NULL dereference when unprovisioning

File: subsys/bluetooth/mesh/access.c
Comment (trond-snekvik) [reply subsys/bluetooth/mesh/access.c:309]: Should this call be removed, now that we have the reset callback in the config server?

Comment (jhedberg) [reply subsys/bluetooth/mesh/access.c:309]: Yes, I think so

File: subsys/bluetooth/mesh/access.c
Comment (jhedberg) [inline subsys/bluetooth/mesh/access.c:346]: I'm wondering if we should do here the reverse of what comp_provision() does, i.e. iterate all elements and set their addresses to UNASSIGNED as well?

Comment (jhedberg) [reply subsys/bluetooth/mesh/access.c:346]: That said, I guess the code wasn't doing it so far (only making the per-model calls), so that can be left as an independent (later) fix

**Review by jhedberg — APPROVED** [review]

File: subsys/bluetooth/mesh/access.c
Comment (trond-snekvik) [reply subsys/bluetooth/mesh/access.c:346]: Good point, yes. I think there are several parts of the unprovision procedure that are a bit fuzzy, and I'm not sure I improved it by doing this. 

After removing this call, we're no longer stopping periodic publishing, but before, we would reinitialize publish's k_delayed_work, which breaks the global timeout queue by wiping the timeout dlist node.

Comment (jhedberg) [reply subsys/bluetooth/mesh/access.c:346]: Seems like something to be fixed in the config server’s reset callback? Anyway, I guess we should hold off on merging this if it’s causing regressions (even though the current code isn’t completely bug-free either)

**Review by LingaoM — APPROVED** [review]

File: subsys/bluetooth/mesh/access.c
Comment (Kupeo) [reply subsys/bluetooth/mesh/access.c:346]: @trond-snekvik 

> Good point, yes. I think there are several parts of the unprovision procedure that are a bit fuzzy, and I'm not sure I improved it by doing this.
> 
> After removing this call, we're no longer stopping periodic publishing, but before, we would reinitialize publish's k_delayed_work, which breaks the global timeout queue by wiping the timeout dlist node.

Sure, reinitialize timer indeed is a problem, it's really an unexpected discovery. 

but, what mean `we're no longer stopping periodic publishing` ??? , i'm find k_delayed_work already canceled in `b_mesh_reset -> bt_mesh_subnet_del -> bt_mesh_app_key_del -> mod_unbind `func.

Comment (Kupeo) [reply subsys/bluetooth/mesh/access.c:346]: > I'm wondering if we should do here the reverse of what comp_provision() does, i.e. iterate all elements and set their addresses to UNASSIGNED as well?

I agree, it can be necessary to do this in another commit.

Comment (trond-snekvik) [reply subsys/bluetooth/mesh/access.c:346]: > i'm find k_delayed_work already canceled in `b_mesh_reset -> bt_mesh_subnet_del -> bt_mesh_app_key_del -> mod_unbind` func.

Well spotted, you're right! I missed that.

So if this isn't a regression after all, I think it could be safe to merge this after the release.

**Pull request merged**
